iso9660/iso9660fs.8
.Dd April 26, 2025
.Dt ISO9660FS 8
.Os
.Sh NAME
.Nm iso9660fs
.Nd iso 9660 filesystem driver
.Sh SYNOPSIS
.Nm
.Op Fl bf
.Op Fl o Ar mount-options
.Op Fl p Ar path
.Ar device
.Ar mountpoint
.Sh DESCRIPTION
.Nm
mounts the ISO 9660 aka ECMA-119 filesystem on the
.Ar device
for reading at the
.Ar mountpoint .
.Nm
forks and runs in the background once the filesystem has been mounted.
.Pp
A compatible ISO 9660 filesystem can be created using
.Xr xorriso 1 .
Filesystems can be automatically mounted by
.Xr init 8
on boot by creating an entry in
.Xr fstab 5
with the
.Sy iso9660
type.
.Pp
.Nm
implements the System Use Sharing Protocol that allows extensions such as Rock
Ridge.
.Pp
.Nm
implements the Rock Ridge 1.10 and 1.12 extensions that provide POSIX metadata,
such as long file names with arbitrary characters, deep directory hierarchies,
permissions, owner and group metadata, special files, hard links, symbolic
links, multiple timestamps with 10 ms precision until year 9999, and more.
Sparse files are not implemented.
.Pp
The options are as follows:
.Bl -tag -width "12345678"
.It Fl b , \-background
Run in the background in a subprocess after successfully mounting.
This is the default behavior.
.It Fl f , \-foreground
Run in the foreground and do not detach in a subprocess.
.It Fl o Ar mount-options
Mount the filesystem with the following comma-separated
.Ar mount-options :
.Bl -tag -width "1234567890"
.It Sy cache Ns = Ns Ar size
Sets the filesystem cache to the specified
.Ar size ,
which is measured in an optional suffix:
.Sy %
for percent of system memory,
.Sy K
for KiB,
.Sy M
for MiB,
or
.Sy G
for GiB.
The default is
.Sy 10%
of the system memory.
.It Sy nosusp
Disable the System Use Share Protocol and therefore Rock Ridge.
.It Sy norock
Disable the Rock Ridge extensions.
.It Sy ro
Mount the filesystem as read-only.
This option is the default behavior and write support is not possible.
.El
.It Fl p , \-pretend-mount-path Ns = Ns Ar path
When answering requests about where the filesystem is mounted, reply with this
.Pa path
instead of the actual mountpoint in
.Xr tcgetblob 2 .
This behavior is useful for
.Xr chroot 2
environments.
.El
.Pp
The following identifiers are available to recognize a filesystem in
.Xr fstab 5
in the
.Ft fs_spec
field and can be queried with the
.Xr disked 8
.Sy id
command:
.Bl -tag -width "12345678"
.It Sy LABEL Ns = Ns Ar label
Search for a filesystem with this
.Ar label
as the volume identifier.
.It Sy ABSTRACT_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the abstract file id.
.It Sy APPLICATION_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the application id.
.It Sy BIBLIOGRAPHIC_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the bibliographic file id.
.It Sy COPYRIGHT_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the copyright file id.
.It Sy DATA_PREPARER_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the data preparer id.
.It Sy PUBLISHER_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the publisher id.
.It Sy VOLUME_SET_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the volume set id.
This identifier may be a good choice for the filesystem search if it has been
set to a unique value.
.It Sy SYSTEM_ID Ns = Ns Ar value
Search for a filesystem with this
.Ar value
as the system id that can interpret the boot data.
.It Sy UUID Ns = Ns Ar creation-time
Search for a filesystem with this
.Ar creation-time
in the YYYY-MM-DD-HH-MM-SS-CC format, without respecting the time zone field.
.Ar uuid .
This identifier is generally recommended for filesystem searches to avoid
mounting the wrong filesystem, but may not be unique, and another identifier may
be better suited depending on the volume.
.El
.Sh IMPLEMENTATION NOTES
If
.Fl f ,
then
.Nm
signals readiness when it has successfully mounted the filesystem.
.Pp
The System Use Sharing Protocol continuation areas are limited to 32 blocks.
.Sh ASYNCHRONOUS EVENTS
.Bl -tag -width "SIGTERM"
.It Dv SIGTERM
Request daemon termination.
.Nm
will exit after gracefully unmounting the filesystem.
.El
.Sh EXIT STATUS
If
.Fl b ,
.Nm
exits 0 after successfully mounting the filesystem and serving requests in a
background process.
.Pp
If
.Fl f ,
.Nm
signals readiness on the
.Ev READYFD
file descriptor when filesystem has been mounted, and continues running in the
same process.
.Pp
.Nm
runs as a
.Xr daemon 7
until stopped by
.Dv SIGTERM
or until the filesystem is unmounted with
.Xr unmount 8
or
.Xr unmount 2 ,
after which
.Nm
exits 0 if the filesystem was cleanly unmounted.
.Pp
.Nm
exits non-zero on fatal errors.
.Sh EXAMPLES
.Bd -literal
$ iso9660fs /dev/foo0 /mnt
$ ls /mnt
$ unmount /mnt
.Ed
.Sh SEE ALSO
.Xr xorriso 1 ,
.Xr fstab 5 ,
.Xr disked 8 ,
.Xr init 8 ,
.Xr unmount 8
.Sh STANDARDS
.Rs
.%D 2023
.%I ISO International Organization for Standardization
.%R ISO/IEC 9660:2023
.%T Volume and File Structure of CD-ROM for Information Interchange
.%Q ISO/IEC JTC 1/SC 23
.Re
.Pp
.Rs
.%D December 2024
.%I Ecma International
.%R ECMA-119 5th Edition
.%T Volume and File Structure of CDROM for Information Interchange
.Re
.Pp
.Rs
.%D July 8, 1994
.%I Institute of Electrical and Electronics Engineers
.%Q IEEE CD-ROM File System Format Working Group
.%R IEEE P1281 1.12
.%T System Use Sharing Protocol
.Re
.Pp
.Rs
.%D July 8, 1994
.%I Institute of Electrical and Electronics Engineers
.%Q IEEE CD-ROM File System Format Working Group
.%R IEEE P1282 1.12
.%T Rock Ridge Interchange Protocol
.Re
.Sh BUGS
Hard links are not perfectly implemented, as the different directory entries
have different inode numbers due to current technical limitations.
.Pp
Multiple sessions are not supported yet due to no driver support.
.Pp
Multiple file extents are not supported.
.Pp
Rock Ridge sparse files are not implemented.
This lack of support matches
.Xr xorriso 1 .
.Pp
Interleaved files via the Interleave Gap Size and File Unit Size are not
implemented.
This lack of support matches
.Xr xorriso 1 .
.Pp
.Xr readdir 2
runs in O(n) linear time, which takes O(n^2) quadratic time when reading a full
directory.
.Pp
ISO 9660 Extended Attributes are not implemented, as Rock Ridge provides the
needed features.
.Pp
The Rock Ridge long form timestamps have a four digit year field and will
overflow in the year 10000.
If this extension is not used, the timestamps use a byte count since the year
1970 and will overflow in the year 2156.
.Pp
The owner and group values are limited to 32-bit.
.Pp
The logical sector size must be 2048 bytes.
Supporting other sector sizes would require searching multiple offsets for the
sector size used by the filesystem, which might not match the underlying
device.


iso9660/iso9660fs.cpp
/*
 * Copyright (c) 2013-2016, 2022-2023, 2025 Jonas 'Sortie' Termansen.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * iso9660fs.cpp
 * Implementation of the ISO 9660 filesystem.
 */

#include <sys/stat.h>
#include <sys/types.h>

#include <assert.h>
#include <dirent.h>
#include <endian.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <inttypes.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#if defined(__sortix__)
#include "fsmarshall.h"
#else
#include "fuse.h"
#endif

#include "block.h"
#include "device.h"
#include "filesystem.h"
#include "inode.h"
#include "ioleast.h"
#include "iso9660.h"
#include "iso9660fs.h"
#include "util.h"

uid_t request_uid;
uid_t request_gid;

mode_t HostModeFromFsMode(uint32_t mode)
{
	mode_t hostmode = mode & 07777;
	if ( ISO9660_S_ISSOCK(mode) ) hostmode |= S_IFSOCK;
	if ( ISO9660_S_ISLNK(mode) ) hostmode |= S_IFLNK;
	if ( ISO9660_S_ISREG(mode) ) hostmode |= S_IFREG;
	if ( ISO9660_S_ISBLK(mode) ) hostmode |= S_IFBLK;
	if ( ISO9660_S_ISDIR(mode) ) hostmode |= S_IFDIR;
	if ( ISO9660_S_ISCHR(mode) ) hostmode |= S_IFCHR;
	if ( ISO9660_S_ISFIFO(mode) ) hostmode |= S_IFIFO;
	return hostmode;
}

uint32_t FsModeFromHostMode(mode_t hostmode)
{
	uint32_t mode = hostmode & 07777;
	if ( S_ISSOCK(hostmode) ) mode |= ISO9660_S_IFSOCK;
	if ( S_ISLNK(hostmode) ) mode |= ISO9660_S_IFLNK;
	if ( S_ISREG(hostmode) ) mode |= ISO9660_S_IFREG;
	if ( S_ISBLK(hostmode) ) mode |= ISO9660_S_IFBLK;
	if ( S_ISDIR(hostmode) ) mode |= ISO9660_S_IFDIR;
	if ( S_ISCHR(hostmode) ) mode |= ISO9660_S_IFCHR;
	if ( S_ISFIFO(hostmode) ) mode |= ISO9660_S_IFIFO;
	return mode;
}

uint8_t HostDTFromFsDT(uint8_t fsdt)
{
	switch ( fsdt )
	{
	case ISO9660_FT_UNKNOWN: return DT_UNKNOWN;
	case ISO9660_FT_REG_FILE: return DT_REG;
	case ISO9660_FT_DIR: return DT_DIR;
	case ISO9660_FT_CHRDEV: return DT_CHR;
	case ISO9660_FT_BLKDEV: return DT_BLK;
	case ISO9660_FT_FIFO: return DT_FIFO;
	case ISO9660_FT_SOCK: return DT_SOCK;
	case ISO9660_FT_SYMLINK: return DT_LNK;
	}
	return DT_UNKNOWN;
}

void StatInode(Inode* inode, struct stat* st)
{
	memset(st, 0, sizeof(*st));
	st->st_ino = inode->inode_id;
	st->st_mode = HostModeFromFsMode(inode->Mode());
	st->st_nlink = inode->nlink;
	st->st_uid = inode->uid;
	st->st_gid = inode->gid;
	st->st_size = inode->size;
	st->st_atim = inode->atim;
	st->st_ctim = inode->ctim;
	st->st_mtim = inode->mtim;
	st->st_blksize = inode->filesystem->block_size;
	st->st_blocks = divup(st->st_size, (off_t) 512);
}

int main(int argc, char* argv[])
{
	size_t memory;
#ifdef __sortix__
	memstat(NULL, &memory);
#else
	memory = (uintmax_t) sysconf(_SC_PAGE_SIZE) *
	         (uintmax_t) sysconf(_SC_PHYS_PAGES);
#endif
	size_t cache_size = (memory / 10);

	const char* fuse_options = NULL;
	const char* pretend_mount_path = NULL;
	bool foreground = false;
	bool no_rock = false;
	bool no_susp = false;
	enum
	{
		OPT_FUSE_OPTIONS = 257,
	};
	const struct option longopts[] =
	{
		{"fuse-options", required_argument, NULL, OPT_FUSE_OPTIONS},
		{"background", no_argument, NULL, 'b'},
		{"foreground", no_argument, NULL, 'f'},
		{"pretend-mount-path", required_argument, NULL, 'p'},
		{0, 0, 0, 0}
	};
	const char* opts = "bfo:p:";
	int opt;
	while ( (opt = getopt_long(argc, argv, opts, longopts, NULL)) != -1 )
	{
		switch ( opt )
		{
		case OPT_FUSE_OPTIONS: fuse_options = optarg; break;
		case 'b': foreground = false; break;
		case 'f': foreground = true; break;
		case 'o':
		{
			char* arg = optarg;
			char* save;
			char* tok;
			while ( (tok = strtok_r(arg, ",", &save)) )
			{
				if ( !strcmp(tok, "ro") )
					;
				else if ( !strcmp(tok, "rw") )
					errx(1, "-o rw: filesystem is not writable");
				else if ( !strncmp(tok, "cache=", strlen("cache=")) )
				{
					char* end;
					uintmax_t val = strtoumax(tok + strlen("cache="), &end, 10);
					uintmax_t mult;
					if ( !strcmp(end, "%" ) )
						mult = (memory / 100);
					else if ( !strcmp(end, "K" ) )
						mult = 1024ULL << 0;
					else if ( !strcmp(end, "M" ) )
						mult = 1024ULL << 10;
					else if ( !strcmp(end, "G" ) )
						mult = 1024ULL << 20;
					else if ( !strcmp(end, "" ) )
						mult = 1;
					else
						errx(1, "warning: invalid cache size: %s", tok);
					if ( __builtin_mul_overflow(val, mult, &cache_size) )
						errx(1, "warning: invalid cache size: %s", tok);
				}
				else if ( !strcmp(tok, "norock") )
					no_rock = true;
				else if ( !strcmp(tok, "nosusp") )
					no_susp = true;
				else
					warnx("warning: unknown mount option: %s", tok);
				arg = NULL;
			}
			break;
		}
		case 'p': pretend_mount_path = optarg; break;
		default: return 1;
		}
	}

	if ( argc - optind < 1 )
		errx(1, "expected device");
	if ( argc - optind < 2 )
		errx(1, "expected mountpoint");

	const char* device_path = argv[optind + 0];
	const char* mount_path = argv[optind + 1];

	if ( !pretend_mount_path )
		pretend_mount_path = mount_path;

	int fd = open(device_path, O_RDONLY);
	if ( fd < 0 )
		err(1, "%s", device_path);

	struct stat st;
	if ( fstat(fd, &st) < 0 )
		err(1, "stat: %s", device_path);

	blksize_t sector_size = 2048;
	blksize_t logical_sector_size = sector_size < 2048 ? 2048 : sector_size;

	struct iso9660_pvd pvd;
	off_t pvd_offset = (off_t) logical_sector_size * 16;
	if ( preadall(fd, &pvd, sizeof(pvd), pvd_offset) != sizeof(pvd) )
	{
		if ( errno == EEOF )
			errx(1, "Not a valid ISO 9660 filesystem: %s", device_path);
		else
			err(1, "read: %s", device_path);
	}
	if ( memcmp(pvd.standard_identifier, "CD001", 5) != 0 )
		errx(1, "Not a valid ISO 9660 filesystem: %s", device_path);
	if ( pvd.type != TYPE_PRIMARY_VOLUME_DESCRIPTOR )
		errx(1, "Not a valid ISO 9660 filesystem: %s", device_path);
	if ( pvd.version != 1 || pvd.file_structure_version != 1 )
		errx(1, "Unsupported ISO 9660 filesystem version: %s", device_path);
	blksize_t block_size = le32toh(pvd.logical_block_size_le);
	if ( block_size < 512 || logical_sector_size < block_size ||
	     logical_sector_size % block_size )
		errx(1, "Unsupported ISO 9660 block size: %s: %ju", device_path,
		     (uintmax_t) block_size);
	size_t block_limit = cache_size / block_size;

	Device* dev = new Device(fd, device_path, block_size, block_limit);
	if ( !dev )
		err(1, "malloc");
	Filesystem* fs = new Filesystem(dev, pretend_mount_path, &pvd);
	if ( !fs )
		err(1, "%s", device_path);
	fs->no_rock = no_rock;
	fs->no_susp = no_susp;

	uint32_t root_lba;
	uint32_t root_size;
	memcpy(&root_lba, pvd.root_dirent + 2, sizeof(root_lba));
	memcpy(&root_size, pvd.root_dirent + 10, sizeof(root_size));
	fs->root_ino = (iso9660_ino_t) root_lba * (iso9660_ino_t) block_size;

	Inode* root = fs->GetInode(fs->root_ino);
	if ( !root )
		err(1, "GetInode");
	if ( !root->ActivateExtensions() )
		err(1, "ActivateExtensions");
	root->Unref();

	if ( !mount_path )
		return 0;

#if defined(__sortix__)
	(void) fuse_options;
	return fsmarshall_main(argv[0], mount_path, foreground, fs, dev);
#else
	return iso9660_fuse_main(argv[0], mount_path, fuse_options, foreground, fs,
	                         dev);
#endif
}
